add definitions/defaults for GFDL fv3 cubed sphere grids C24, C48, C96, C192, C384

jtruesdal opened the pull request:
### Description of changes
I'm adding the GFDL FV3 dycore to CAM and need to update CTSM with new default boundary data and fv3 grid definitions.  We are finishing up a 3-year project with NOAA to get this dycore into CESM.

Contributors other than yourself, if any: @ekluzek 

CTSM Issues Fixed (include github issue #): #1025

Are answers expected to change (and if so in what way)?
No

Any User Interface Changes (namelist or namelist defaults changes)?
Additions to namelist defaults only for new FV3 grid datasets.  All existing defaults are unchanged.

Testing performed, if any:  I've been running AMIP and fully coupled controls with the C48,C96 datasets for the past two years.  All the other resolutions were created at the same time.
Minimal CTSM testing for now.  Given the null impact of the change on existing CTSM functionality, I would like to know what testing you would like me to perform.

I ran the two sample tests listed in the CTSM wiki just to make sure I didn't make an xml syntax error in my branch.

  ERS_D_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-default (Overall: PASS)
  SMS_D_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-default (Overall: PASS)

**NOTE: Be sure to check your Coding style against the standard:**
https://github.com/ESCOMP/ctsm/wiki/CTSM-coding-guidelines

billsacks commented:
I'm going to defer to @ekluzek on this review.

jtruesdal commented:
Thanks Bill.  Hi Erik.  I'm new to CTSM PRs and don't know your scheduling procedure.  I'm will amend or add this request to any other planning docs you might have.  I checked briefly with Keith Oleson who suggested running a PR by you guys to give you a chance to look at the changes.  Since the changes are limited and don't change answers he was thinking that you might be able to add this in on one of your upcoming commits. Just FYI I'm also making a cime PR with the LND2XXX and XXX2LND mapping files.

ekluzek commented:
OK, I've done some looking over for this PR. There are some changes I need before it comes in. But, also I just have some questions. I also run the namelist unit tester on it, as well as the tools tester and they both work (I did have a small update to one that I pushed to the branch).

First, what do these grids roughly correspond to in terms of resolution? Is C384 higher resolution than C24?

jtruesdal commented:
First off, thanks for your careful look at the code.  

The nomenclature for the new grid is similar to the SE cubed sphere.  The number corresponds to the number of points on a cube face in the x or y direction, 4 faces circle the globe and C96 would provide 384 points around the globe which is just about 1 degree.
C24 ~4 degree
C48  ~2 degree
C96  ~1 degree
C192 ~0.5 degree
C384 ~ 0.25 degree
These seemed to be a minimal set that should cover quick testing at coarse resolutions to workhorse 1 degree and then high-resolution runs.

I'm happy to recreate the files using  release-clm5.0.30a to get the bugfix and put them in a versioned subdirectory.  Do I have permission to create the release-clm5.0.30 subdirectory that these will reside under or does someone in the land group need to do that?

Cutting down on the number of files to be carried around would be preferable. If its already in place in the  release-clm5.0.30 code then I'm happy to use the release tools to create the files and arrange the defaults file in a way that allows me to make use of that functionality. 

Just to make sure I understand, it sounds like I no longer need to create files using the -no-crop option, and just produce the 78pft version.   I use the following command for those types of files, is this correct?

qsub -A P03010039 "./mksurfdata.pl -res usrspec -usr_gname ${dstgrid} -usr_gdate ${datestr}  -years 1850,2000,1850-2000"

These files will no longer have _irrig as part of their name and can be used with/without irrigation.

Lastly, I've been creating some new CESM/CAM tests for FV3 and one of the tests failed because I didn't add in the landuse.timeseries defaults that are also created by the above command.  I would like to add these to the PR following all the caveats mentioned above.

ekluzek commented:
@jtruesdal you should have permission to create a subdirectory on cheyenne. But, I went ahead and created one for you anyway.

And yes the command you gave for mksurfdata.pl is what you need to do. All files will be for 78pfts and none will have "Irrig" in the names. And including 1850-2000 will also create the landuse.timeseries files you need as well.

Note, I want you to use release-clm5.0.30 to create the files, but we want this PR to come to CTSM master. Once, you have the files in place I help you with the changes that need to go in, that will turn on the additional namelist settings so that you can use the 78pft surface dataset files for non-crop simulations.

ekluzek commented:
@jtruesdal we had a meeting last week to talk about CAM grid support in CLM. It was stated that at least C96 would be an officially supported grid. As such we should also include the SCRIP grid and mapping files for it in this PR.

And we should include pointing to at least the SCRIP grid files for the other grids.

jtruesdal commented:
@ekluzek where should the SCRIP grid files go?  Is there an official location? I will make sure the mapping files are upped to the data repo as well as CISL and CGD online filesystems.  Glad to hear that 1 degree will be officially supported.  Can you tell me what changes are needed to the code to satisfy official support?

ekluzek commented:
I just forwarded you a message that should help with some of your questions. I just need the SCRIP grid files somewhere under inputdata. Some of the other grids are under atm/cam, CLM specific ones are under lnd/clm2/mapping/grids. These should probably reside somewhere CAM specific.

In XML these files go into the namelist_defaults_ctsm_tools.xml file.

In terms of official support I need the SCRIP grid files and all the mapping files for C96 (mapping files go into namelist_defaults_ctsm.xml). What official support for a grid means is that when we create a new batch of surface datasets (because we did some update to them) we will update that grid as well. For experimental grids -- we won't automatically regenerate them -- but let the people who are interested in them do so.

ekluzek commented:
@billsacks notes two other things we need to do for a supported grid:

(1) Add a processor layout for it in CTSM's config_pes.xml (important for anything other than low-resolution grids for testing)
(2) Add a test for it to the aux_clm test list

jtruesdal commented:
@ekluzek I've made a first cut at updates to the surfdata files.  I didn't quite know what to do with the SSP entries please take a look at those.  I've removed all the no_crop datasets as requested, placed the fv3 grid descriptions into our CAM directory and put pointers in namelist_defaults_ctsm_tools.xml.  Next up is a pe layout and aux_clm test.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -665,15 +655,15 @@ lnd/clm2/surfdata_map/surfdata_1x1_brazil_16pfts_Irrig_CMIP6_simyr2000_c171214.n
 <!-- Crop surface datasets -->
 <fsurdat hgrid="C384"  sim_year="2000" use_crop=".true." >
-lnd/clm2/surfdata_map/surfdata_C384_78pfts_CMIP6_simyr2000_c181018.nc</fsurdat>
+lnd/clm2/surfdata_map/surfdata_C384_78pfts_CMIP6_simyr2000_c200317.nc</fsurdat>
```
Put these under the release-clm5.0.30 subdirectory as well, like the others. Looks like the filename isn't quite right either as it's missing "_hist_" before _78pfts.

jtruesdal replied:
Done. Fixed file path to reflect Erik's comments.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -783,6 +763,19 @@ lnd/clm2/surfdata_map/surfdata_ne30np4_78pfts_CMIP6_simyr1850_c170824.nc</fsurda
 <fsurdat hgrid="ne120np4"     sim_year="1850" use_crop=".true." >
 lnd/clm2/surfdata_map/surfdata_ne120np4_78pfts_CMIP6_simyr1850_c170824.nc</fsurdat>
 
+<!-- SSP-RCP scenarios -->
```
Actually remove this set. We don't need to have a separate 1850 fsurdat file for the SSP's from the historical one. It's good to run cprnc on the two files to verify that they aren't different though. They should be identical.

jtruesdal replied:
Removed the 1850 surfdat files and associated comments.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -934,6 +938,17 @@ lnd/clm2/surfdata_map/surfdata_ne120np4_78pfts_CMIP6_simyr1850_c170824.nc</fsurd
 <flanduse_timeseries hgrid="ne120np4"      rcp="8.5" sim_year_range="1850-2100" 
  use_crop=".true."  >lnd/clm2/surfdata_map/landuse.timeseries_ne120np4_hist_78pfts_CMIP6_simyr1850-2015_c170824.nc</flanduse_timeseries>
 
+<flanduse_timeseries hgrid="C24"      rcp="8.5" sim_year_range="1850-2100"
```
Same remove these...

jtruesdal replied:
So no 1850-2100 SSP files then?

ekluzek replied:
Whoops, sorry I thought these were another set of fsurdat files.

I'd like to include one SSP5-8.5 landuse timeseries for the C96 case and also point to the same file for the historical case.

For the other files, we want to limit our support for landuse.timeseries files since they are large. So the low resolution files are probably fine. The high resolution cases might be large enough that we don't really want to make them, unless you know they are absolutely needed. And other than the C96 landuse.timeseries, we won't automatically create them when we update surface datasets, but ask you to do so.

jtruesdal replied:
@ekluzek I got rid of everything except the 1 degree and a low resolution (C24 - 4 degree) dataset.

ekluzek replied:
Perfect.

ekluzek commented:
OK, awesome. You are pointing to the SCRIP grid files and the needed mapping files now. And new datasets are made using release-clm5.0.30 and in the release-clm5.0.30 subdirectory. I pointed out a couple changes to make in the commit.

jtruesdal commented:
@ekluzek Can you point me to the file(s) I need to modify to add a clm test?  Any suggestions for a test?

jtruesdal commented:
@ekluzek I added test ERI_D_Ld9.C96_C96_mg17.I2000Clm50BgcCruGs.cheyenne_intel to the aux_clm test list.
Since this test was requiring a no_crop and irrigation dataset and remembering that you mentioned that the new datasets would be used for both cases, I removed the crop and irrigation checks from the new fv3 srfdat files.  I also needed to update the cime tag to a more recent version in my sandbox to get the test to run.  It has updated fv3 resolution and grid data.  I guess until the cime tag is updated on master we should comment out/remove the added test?

Removing the crop and irrigation checks for the new fv3 srfdata files was just a guess.  I don't know what the proper solution is.

jt

ekluzek commented:
@jtruesdal can you grant me push access to your fork? Go to your fork, "settings", "manage access" and there's a button something like "add collaborator". Then it goes to me and I'll accept it.

I'll help you finish the namelist settings. I need to figure it out anyway. So I'll figure it out and then push the changes to your fork/branch.

Thanks.

jtruesdal commented:
@ekluzek I did add you as a collaborator.  I was wondering if you now have access?

ekluzek commented:
@jtruesdal yes, I got push access to your branch/fork, and I've pushed a couple of changes to this PR. There are problems with the cime version that also need to be fixed here. I suppose I'll send those to cime and CC you on them.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -794,6 +814,11 @@ lnd/clm2/surfdata_map/surfdata_ne120np4_78pfts_CMIP6_simyr1850_c170824.nc</fsurd
  use_crop=".true."   >lnd/clm2/surfdata_map/landuse.timeseries_1x1_numaIA_hist_78pfts_CMIP6_simyr1850-2015_c170917.nc</flanduse_timeseries>
 
+<flanduse_timeseries hgrid="C24"      sim_year_range="1850-2000"
+ use_crop=".true."  >lnd/clm2/surfdata_map/release-clm5.0.30/landuse.timeseries_C24_SSP5-8.5_78pfts_CMIP6_simyr1850-2100_c200317.nc</flanduse_timeseries>
```
use_crop setting should be removed here, so that it can be used for both use_crop=".true. and false. Same for C96 as well. You should also add a compset that can test it, and make sure it works.

ekluzek replied:
@jtruesdal I think this might be my last change to resolve.

ekluzek replied:
- [ ] Resolved

jtruesdal replied:
OK I got rid of the use_crop settings for C24 and C96 in the defaults file. Thanks for the reminder.  

I will need help setting up the compset.  I have no idea how the nomenclature of the alias IHistClm50SpGs should map into the long name. For instance how should it differ from IHistClm50SpG.

I'm happy to run a test but will need help setting it up and identifying what to look at to make sure the results are valid.  Or is it just a smoke test to make sure it runs?

ekluzek replied:
Now that the cime version is updated the compsets aren't as picky anymore. We had to remove the Glacier model previously, because there weren't mapping files for it. Now, there is, so a general compset should work.

ekluzek commented:
@jtruesdal add a IHistClm50SpGs compset so the landuse time series files can be tested.

Also what is the status of getting the C24, C48, C192, C384 grids into cime? They are missing in this version of cime. It looks like there are domain files that could be added for them.

jtruesdal commented:
Thanks.  The cime grids were added in a PR that has been accepted. I think
they are a part of the problematical cime5.8.16 but not sure if they are
also in an earlier tag.

jt

On Wed, Apr 22, 2020 at 9:37 AM Erik Kluzek <notifications@github.com>
wrote:

> @jtruesdal <https://github.com/jtruesdal> add a IHistClm50SpGs compset so
> the landuse time series files can be tested.
>
> Also what is the status of getting the C24, C48, C192, C384 grids into
> cime? They are missing in this version of cime. It looks like there are
> domain files that could be added for them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CTSM/pull/891#issuecomment-617855410>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACG3NX3MKNFBCLLXJBK5FJDRN4FMLANCNFSM4KJZRAUQ>
> .
>

ekluzek commented:
cime issues ESMCI/cime/#3495, ESMCI/cime/#3497, ESMCI/cime/#3498 should be resolved along with this PR.

ekluzek commented:
> Thanks. The cime grids were added in a PR that has been accepted. I think they are a part of the problematical cime5.8.16 but not sure if they are also in an earlier tag. jt

Hmmm. I didn't see them on latest master. I created a couple issues regarding this in cime and assigned it to you. Can you look into that?

ekluzek commented:
OK, the cime issues other than the first one are all addressed in a cime PR by @jtruesdal. We just need this PR to point to a new version of cime that includes that update, when that gets into cime. Since, updating cime is often problematic we might want to do that update outside of this PR. My reading of the cime PR is that they have been accepted -- but NOT merged to master yet, and that's why they aren't in a cime tag yet, nor on master.

jtruesdal commented:
Maybe I'm looking in the wrong place, I did see them in master.  This was
one of the files in the PR that was merged.
https://github.com/ESMCI/cime/blob/master/config/cesm/config_grids.xml


On Wed, Apr 22, 2020 at 12:18 PM Erik Kluzek <notifications@github.com>
wrote:

> OK, the cime issues other than the first one are all addressed in a cime
> PR by @jtruesdal <https://github.com/jtruesdal>. We just need this PR to
> point to a new version of cime that includes that update, when that gets
> into cime. Since, updating cime is often problematic we might want to do
> that update outside of this PR. My reading of the cime PR is that they have
> been accepted -- but NOT merged to master yet, and that's why they aren't
> in a cime tag yet, nor on master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CTSM/pull/891#issuecomment-617946917>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACG3NX5OFPUZMFZJ6TMT5UDRN4YIFANCNFSM4KJZRAUQ>
> .
>

ekluzek commented:
OK, yes you are right. They are just not in this version of cime for this version of CTSM. I remember checking, but not seeing it -- but must have been looking at the wrong thing. Actually, you should be able to just merge to latest ctsm ctsm1.0.dev089 and get a version of CIME that should work (since it's using cime5.8.17). Could you do that?

Then there would just be a few things in CTSM to address.

jtruesdal commented:
Sure. I will try that and get back to you.


On Wed, Apr 22, 2020 at 12:38 PM Erik Kluzek <notifications@github.com>
wrote:

> OK, yes you are right. They are just not in this version of cime for this
> version of CTSM. I remember checking, but not seeing it -- but must have
> been looking at the wrong thing. Actually, you should be able to just merge
> to latest ctsm ctsm1.0.dev089 and get a version of CIME that should work
> (since it's using cime5.8.17). Could you do that?
>
> Then there would just be a few things in CTSM to address.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CTSM/pull/891#issuecomment-617957485>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACG3NX4YNTA4S75RZINN5VTRN42SDANCNFSM4KJZRAUQ>
> .
>

jtruesdal commented:
Ok I merged in latest dev89 mods and made a few successful tests for FV3.  The code was based on the cesm2_2_alpha04d which used the standard cime5.8.17.

ekluzek commented:
@jtruesdal there's one change that I see that I requested for you to change in the namelist defaults file and then try a test for it. Then this should be ready for me to take over and run the full testing with it.

One other change that I'll make is to make C96 a supported grid that we automatically make surface datasets for.

ekluzek commented:
Oh, and @jtruesdal one other question for you. The FV3 datasets add up to a 1TByte of disk storage, and I noticed that there were some landuse.timeseries files for T384 a high resolution grid. I'm wondering if we really need those high resolution files?

I'm thinking the way I might handle it is to NOT check them into inputdata, and just leave the copy in inputdata on cheyenne. They could be restored from backup still, but you wouldn't be able to easily get them on another machine like you can with files that are in inputdata.

jtruesdal commented:
Just so I understand, you are proposing not to check in the 0.25 degree (C384) FV3 land data into the data repository but just leave a local copy on Cheyenne. The three other FV3 resolutions 4, 2, and 1 degree would be checked in? 

Are you also now limiting what gets checked in by the other supported dycores? Is it no longer practical to use the inputdata repository for high resolution land data?

ekluzek commented:
@jtruesdal we are concerned about having multiple TBytes of surface dataset info. each time we create a new set of surface datasets. A TByte is doable now, but it's still a lot of data.  By far the biggest part of the data is the landuse.timeseries files. It looks like storing high resolution files for a few time-slices is fine. But, the timeseries files can be the majority of that TByte of space. And actually one limitation of the landuse.timeseries data is that at higher resolutions it's actually lower resolution data that takes up a lot of space because it's stored at a higher resolution. So I think we need to have a better way to store it for very high resolutions.

Right now it takes a TByte of storage for each time we create new datasets. There are also being added some variable resolution grids that add another half TByte, and if these grids add a nother full TByte, that means we have 2.5 TBytes each time we create new supported datasets. And that's storage on glade on cheyenne as well as in svn inputdata. Both of which are finite resources where yes a single TByte is OK at this point, but it only takes a few multiples of that to start running into trouble.

The advice we've given Adam in #994 is to bring in single year surface datasets for any resolution, but try to only have landuse.timeseries files for lower resolutions that they know they plan on doing simulations with.  See especially this note from Adam (https://github.com/ESCOMP/CTSM/issues/994#issuecomment-621365579). We also want to distinguish between grids that are considered "supported" (which means we will recreate new versions anytime we update surface datasets)  and "community" (which means we would have you create new versions as needed when updates that require new surface datasets happen).

Part of what we don't want is to waste storage space on creating datasets that are then not used. But, we also don't want to block important science being done. So we have to make sure everyone is aware of the limitations here.

Since, this is long. It might be best to have a zoom call to go over this. Should we set that up?

jtruesdal commented:
Hi Erik:

Thanks for the background and your comprehensive reply.  Most of what I was
worried about would be taken care of by being able to distinguish between
supported and community grids thus alerting a potential user of whether a
given grid is available via our repository. I like the idea of community
supported and hopefully we will have community space set aside for user
generated data that can be used by everyone.  I'm good with just providing
the lower resolution timeseries data for this commit and will just hold on
to the bigger files to see if they are needed before a new version requires
rebuilding the data.

I will still need some help defining the compset you asked for and making a
validation test of the crop/no crop datasets.

jt

On Mon, May 4, 2020 at 4:08 PM Erik Kluzek <notifications@github.com> wrote:

> @jtruesdal <https://github.com/jtruesdal> we are concerned about having
> multiple TBytes of surface dataset info. each time we create a new set of
> surface datasets. A TByte is doable now, but it's still a lot of data. By
> far the biggest part of the data is the landuse.timeseries files. It looks
> like storing high resolution files for a few time-slices is fine. But, the
> timeseries files can be the majority of that TByte of space. And actually
> one limitation of the landuse.timeseries data is that at higher resolutions
> it's actually lower resolution data that takes up a lot of space because
> it's stored at a higher resolution. So I think we need to have a better way
> to store it for very high resolutions.
>
> Right now it takes a TByte of storage for each time we create new
> datasets. There are also being added some variable resolution grids that
> add another half TByte, and if these grids add a nother full TByte, that
> means we have 2.5 TBytes each time we create new supported datasets. And
> that's storage on glade on cheyenne as well as in svn inputdata. Both of
> which are finite resources where yes a single TByte is OK at this point,
> but it only takes a few multiples of that to start running into trouble.
>
> The advice we've given Adam in #994
> <https://github.com/ESCOMP/CTSM/issues/994> is to bring in single year
> surface datasets for any resolution, but try to only have
> landuse.timeseries files for lower resolutions that they know they plan on
> doing simulations with. See especially this note from Adam (#994 (comment)
> <https://github.com/ESCOMP/CTSM/issues/994#issuecomment-621365579>). We
> also want to distinguish between grids that are considered "supported"
> (which means we will recreate new versions anytime we update surface
> datasets) and "community" (which means we would have you create new
> versions as needed when updates that require new surface datasets happen).
>
> Part of what we don't want is to waste storage space on creating datasets
> that are then not used. But, we also don't want to block important science
> being done. So we have to make sure everyone is aware of the limitations
> here.
>
> Since, this is long. It might be best to have a zoom call to go over this.
> Should we set that up?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CTSM/pull/891#issuecomment-623678814>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACG3NX7DKSYS2NZMHUEQUJ3RP4OFRANCNFSM4KJZRAUQ>
> .
>

jtruesdal commented:
Ok.  Brought everything up to master to get rid of conflicts. But now I'm not sure if I should have done that, is that OK?  I'm still unsure about the IHistClm50SpGs compset and test.  I will add the 2 lower resolution (C24 and C96) timeseries listed here to the repository and make sure the others are available on Cheyenne.  I also wasn't sure about the values for build-namelist_test.pl.  I think you originally modified this but not sure if these values are still good.

ekluzek commented:
I think you should add a test for C96 with IHistClm50BgcCrop and IHistClm50Sp. And just make sure that the test runs. Should we also test. the lower resolution of C24 as well? Should we consider both of them (C24 and C96) supported?

Once, you do those things, it's. probably time for me to take it over. I'll make any changes needed the build-namelist tester, as well as making C96 grids being made by default.

jtruesdal commented:
I'm fine with just C96 support.  C24 is mostly for debug.  I added the two tests you asked for and successfully ran them. I also added the setup for the IHistClm50Sp to collapse the 78pfts to 16 for C96.

ERS_D_Ld10.C96_C96_mg17.IHistClm50Sp.cheyenne_intel.clm-collapse_pfts_78_to_16_decStart_C96 (Overall: PASS) details:

ERP_D_Ld5.C96_C96_mg17.IHistClm50BgcCrop.cheyenne_intel.clm-allActive (Overall: PASS) details:

ekluzek commented:
OK, it's passed testing on cheyenne and izumi.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -768,6 +768,18 @@ lnd/clm2/surfdata_map/release-clm5.0.18/surfdata_1x1_brazil_hist_16pfts_Irrig_CM
+<fsurdat hgrid="C48"  sim_year="2000">
+lnd/clm2/surfdata_map/release-clm5.0.30/surfdata_C48_hist_78pfts_CMIP6_simyr2000_c200317.nc</fsurdat>
+<fsurdat hgrid="C24"  sim_year="2000">
+lnd/clm2/surfdata_map/release-clm5.0.30/surfdata_C24_hist_78pfts_CMIP6_simyr2000_c200317.nc</fsurdat>
```
Whoops. This addition of f05 shouldn't be here...

ekluzek commented:
I'm saving the surface and landuse.timeseries files to svn inpudata. I'm asking @jtruesdal to do that for the mapping files.
